### PR TITLE
 Add ability to set system wide options for slirp4netns

### PR DESCRIPTION
**Pull request by ashley-cui**
Wire in containers.conf options for slirp
Vendor in containers/common v0.30.0 

Closes https://github.com/containers/podman/issues/7888

### CODE REVIEW

**Comment by ashley-cui** [comment]
@containers/podman-maintainers  PTAL

**Comment by ashley-cui** [comment]
failing tests.. jumping back in

**Comment by rhatdan** [comment]
LGTM
/approve

**Comment by rhatdan** [comment]
@AkihiroSuda @tmds PTAL

**Comment by AkihiroSuda** [comment]
port_handler=slirp4netns is incompatible with CNI networks and should be ignored with a warning message

**Comment by tmds** [comment]
> port_handler=slirp4netns is incompatible with CNI networks and should be ignored with a warning message.

What does that mean? Can a non-root user create a network, add containers to it, and allow those containers to access services bound to the host's localhost?

File: test/e2e/config/containers.conf [hunk 1]
```diff
@@ -52,3 +52,7 @@ dns_options=[ "debug", ]
 tz = "Pacific/Honolulu"
 
 umask = "0002"
+
+[engine]
+
+network_cmd_options=["allow_host_loopback=true"]
```
Comment (tmds) [inline test/e2e/config/containers.conf:58]: Will this setting apply also to networks created by non-root users via `podman network create`?

Comment (AkihiroSuda) [reply test/e2e/config/containers.conf:58]: Yes, it should work

**Comment by AkihiroSuda** [comment]
I mean `podman run -p 8080:80 --network some-cni-network` only works with rootlesskit port driver, not with slirp4netns port driver currently

**Comment by tmds** [comment]
I don't know enough about the code to approve the PR, but it looks like it has everything needed.
Thanks for implementing this feature @ashley-cui !

**Comment by TomSweeneyRedHat** [comment]
LGTM

**Comment by umohnani8** [comment]
LGTM but needs a rebase

File: test/e2e/containers_conf_test.go [hunk 2]
```diff
@@ -258,6 +258,12 @@ var _ = Describe("Podman run", func() {
 		Expect(session.OutputToString()).To(Equal("0002"))
 	})
 
+	It("podman set network cmd options slirp options to allow host loopback", func() {
+		session := podmanTest.Podman([]string{"run", "--network", "slirp4netns", ALPINE, "ping", "-c1", "10.0.2.2"})
```
Comment (tmds) [inline test/e2e/containers_conf_test.go:262]: @AkihiroSuda is the ip of `10.0.2.2` for the host, something defined by `slirp4netns` or by `podman`?

Comment (AkihiroSuda) [reply test/e2e/containers_conf_test.go:262]: This is slirp4netns's default value https://github.com/rootless-containers/slirp4netns/blob/master/slirp4netns.1.md#description

**Comment by rhatdan** [comment]
@ashley-cui Rebase so we can get this merged.

**Comment by ashley-cui** [comment]
@rhatdan rebased

**Comment by tmds** [comment]
@ashley-cui I'd like to try this out.

I have installed:
```
podman.x86_64                                          2:2.2.1-1.fc32                         @updates-testing                                               
podman-plugins.x86_64                                  2:2.2.1-1.fc32                         @updates-testing
```
I have edited:
```
$ cat $HOME/.config/containers/containers.conf 
[engine]

network_cmd_options=["allow_host_loopback=true"]
```
Then I did a `reboot`.

However, `slirp4netns` is still running with `--disable-host-loopback`:
```
$ ps aux | grep slirp
tmds        1538  0.0  0.0   4960  3080 ?        S    11:10   0:00 /usr/bin/slirp4netns --disable-host-loopback --mtu 65520 --enable-sandbox --enable-seccomp -c -e 3 -r 4 --netns-type=path /run/user/1000/netns/cni-066a38cb-f880-58a7-4b68-b6f8094d4870 tap0
```

I can't access TCP services running on `localhost`.
Strangely, this doesn't fail:
```
$ podman run --network mynetwork alpine ping -c1 10.0.2.2
PING 10.0.2.2 (10.0.2.2): 56 data bytes
64 bytes from 10.0.2.2: seq=0 ttl=42 time=0.208 ms

--- 10.0.2.2 ping statistics ---
1 packets transmitted, 1 packets received, 0% packet loss
round-trip min/avg/max = 0.208/0.208/0.208 ms
```
That means the test added here doesn't really verify localhost is now accessible.
@AkihiroSuda this may be a bug in `slirp4netns`, or is this expected?

**Comment by AkihiroSuda** [comment]
> @AkihiroSuda this may be a bug in slirp4netns, or is this expected?

As slirp4netns is running with --disable-host-loopback, this behavior is expected from POV of slirp4netns.
Probably a bug on Podman side.

**Comment by tmds** [comment]
I mean: is `ping -c1 10.0.2.2` supposed to work when `slirp4netns` is running with `--disable-host-loopback`?

**Comment by AkihiroSuda** [comment]
Ping is always allowed

**Comment by tmds** [comment]
> Ping is always allowed

That means the test added here will always succeed independent of `allow_host_loopback` being `true` or `false`.

> I have installed:

The version I installed from testing may not have this change yet. I don't know how to check.

**Comment by ashley-cui** [comment]
@tmds it seems to work for me. I built podman from master, since this feature is not in a release yet 

 with network_cmd_options=["allow_host_loopback=true"] : 
```
$ ps aux | grep slirp
acui       14782  0.0  0.0   4640  2764 pts/2    S    09:31   0:00 /usr/bin/slirp4netns --mtu 65520 --enable-sandbox --enable-seccomp -c -e 3 -r 4 --netns-type=path /run/user/1000/netns/cni-5f350209-f934-bf00-56ec-98a3adb06fd1 tap0
```

without network_cmd_options=["allow_host_loopback=true"] : 
```
$ ps aux | grep slirp
acui       14921  0.0  0.0   4640  2716 pts/2    S    09:32   0:00 /usr/bin/slirp4netns --disable-host-loopback --mtu 65520 --enable-sandbox --enable-seccomp -c -e 3 -r 4 --netns-type=path /run/user/1000/netns/cni-751db779-71a2-05bf-dff9-0682fab68a0f tap0
```

**Comment by tmds** [comment]
@ashley-cui is there a daily build I can use?

**Comment by ashley-cui** [comment]
@tmds I don't think so, (?) but you're welcome to build it yourself and test it

**Pull request merged**